### PR TITLE
Integrate RNN mode into web GUI.

### PR DESCRIPTION
**Pull request by rit1217**

### CODE REVIEW

**Comment by rit1217** [comment]
At this point, the model can predict only one sequence(title). I'm still finding a way to generate and score multiple sequences.

File: Dockerfile [hunk 1]
```diff
@@ -25,7 +25,7 @@ RUN pip3 install \
   --no-cache-dir \
   --find-links=/wheels \
   -r requirements.txt
-RUN ["python3", "-m", "model_graph"]
+RUN ["python3", "-m", "model_rnn", "-t"]
```
Comment (IDDT) [inline Dockerfile:28]: I don't think you can train it like that in docker, maybe better copy pretrained weights from `temp/`?

File: Dockerfile [hunk 2]
```diff
@@ -19,13 +19,8 @@ RUN pip3 wheel --wheel-dir=/wheels -r requirements.txt
...
 FROM build AS train
 WORKDIR /train
-COPY . .
-RUN pip3 install \
-  --no-index \
-  --no-cache-dir \
-  --find-links=/wheels \
-  -r requirements.txt
-RUN ["python3", "-m", "model_rnn", "-t"]
+COPY ./temp ./temp
```
Comment (IDDT) [inline Dockerfile:22]: You might have other files in temp, wouldn't want to copy those.

File: model_rnn/adapter.py [hunk 3]
```diff
@@ -7,27 +7,52 @@
...
         return self
 
-    def predict(self, prefix:str='', max_length:int=100) -> list:
+    def predict(self, prefix:str='', top_n:int=10, max_length:int=100) -> list:
+        title = preprocess_text(prefix)[:-1]
+        result = self._predict_helper(title, top_n, max_length)
+        result.sort(key=lambda item: item['score'], reverse=True)
+        return result
+
+    def _predict_helper(self, title:str, top_n:int=10, max_length:int=100, result_list:list=[], prefix_score:float=0) -> list:
```
Comment (IDDT) [inline model_rnn/adapter.py:24]: This part can be done better. Try to re-architect this algo to make it simpler.

File: model_rnn/adapter.py [hunk 4]
```diff
@@ -33,26 +36,16 @@ def _predict_helper(self, title:str, top_n:int=10, max_length:int=100, result_li
...
-            if len(result_list) < top_n :
-                result_list.append(result)
-            #Replace lowest score title with the new result
-            elif score > min(result_list, key=lambda item: item['score'])['score']:
-                min_ix = result_list.index( min(result_list, key=lambda item: item['score']))
-                result_list[min_ix] = result
-                
-            return result_list
+
+            yield {'title':title[1:], 'score':score}
```
Comment (IDDT) [inline model_rnn/adapter.py:51]:
```suggestion
            yield title[1:], score
```
In python it's preferred to return tuples

File: model_rnn/adapter.py [hunk 5]
```diff
@@ -33,26 +36,16 @@ def _predict_helper(self, title:str, top_n:int=10, max_length:int=100, result_li
                 topv, topi = output.reshape(-1).topk(2)
                 top_char = IX_TO_CHAR[topi[0].item()]
                 top_2_char = IX_TO_CHAR[topi[1].item()]
+
+                if n > 0 and top_2_char != EOS:
+                    n = n // 2
+                    for result in self._predict_helper(title + top_2_char, n, max_length, score + topv[1].item()):
```
Comment (IDDT) [inline model_rnn/adapter.py:42]: This approach can be optimized better. Once you vectorize user prefix, you don't ever need to vectorize it again. It's not what you are doing here through.

File: model_rnn/adapter.py [hunk 6]
```diff
@@ -18,34 +18,36 @@ def load(self):
...
+
+    def _predict_helper(self, X:torch.tensor, n:int=20, max_length:int=100, prefix_score:float=0) -> Generator[Tuple, None, None]:        
         score = prefix_score
+        title_len = X.size()[0]
         with torch.no_grad():
-            X = make_input_vect(title)
             hidden = None
-            for i in range(len(title) - 1):
+            for i in range(title_len - 1):
                 output, hidden = self.rnn.predict(X[-1].reshape(1, 1, -1), hidden)
```
Comment (IDDT) [inline model_rnn/adapter.py:35]: Need to traverse through user provided prefix only once.

**Review by IDDT — APPROVED** [review]

**Pull request merged**